src/systemd.c

```diff
@@ -0,0 +1,512 @@
+                    .collectd_type = METRIC_TYPE_COUNTER,
+                },
+                {.collectd_type = METRIC_TYPE_UNTYPED},
+            },
+    },
+    {
+        .accounting_flag = "IPAccounting",
+        .metrics =
+            (systemd_metric[]){
+                {
+                    .name = "IPEgressBytes",
+                    .dbus_type = "t",
+                    .collectd_type = METRIC_TYPE_COUNTER,
+                },
+                {
+                    .name = "IPEgressPackets",
+                    .dbus_type = "t",
+                    .collectd_type = METRIC_TYPE_COUNTER,
+                },
+                {
+                    .name = "IPIngressBytes",
+                    .dbus_type = "t",
+                    .collectd_type = METRIC_TYPE_COUNTER,
+                },
+                {
+                    .name = "IPIngressPackets",
+                    .dbus_type = "t",
+                    .collectd_type = METRIC_TYPE_COUNTER,
+                },
+                {.collectd_type = METRIC_TYPE_UNTYPED},
+            },
+    },
+    {
+        .accounting_flag = "TasksAccounting",
+        .metrics =
+            (systemd_metric[]){
+                {
+                    .name = "TasksCurrent",
+                    .dbus_type = "t",
+                    .collectd_type = METRIC_TYPE_GAUGE,
+                },
+                {.collectd_type = METRIC_TYPE_UNTYPED},
+            },
+    },
+};
+
+typedef struct {
+  char *path;
+  bool is_slice;
+} unit;
+
+static unit *units = NULL;
+
+static size_t units_num = 0;
+
+static sd_bus *bus = NULL;
+
+static bool was_service = false;
+static bool was_slice = false;
+
+static int introspect_prop(xmlXPathContextPtr xpath_ctx,
+                           char const interface[static 1], char **prop) {
+  char query[128];
+  sprintf(query,
+          "//interface[@name=\"%s\"]/"
+          "property[@name=\"%s\"]",
+          interface, *prop);
+  xmlXPathObjectPtr xpath_obj =
+      xmlXPathEvalExpression(BAD_CAST(query), xpath_ctx);
+  if (xpath_obj == NULL) {
+    ERROR("Can't get xpath object");
+    return EXIT_FAILURE;
+  }
+  if (xpath_obj->nodesetval->nodeNr == 0) {
+    WARNING("This systemd version doesn't provide %s property in %s interface",
+            *prop, interface);
+    *prop = NULL;
+  }
+  return EXIT_SUCCESS;
+}
+
+static int introspect_unit(unit *unit) {
+  sd_bus_error sd_bus_err = SD_BUS_ERROR_NULL;
+  sd_bus_message *m = NULL;
+  int r = sd_bus_call_method(bus, "org.freedesktop.systemd1", unit->path,
+                             "org.freedesktop.DBus.Introspectable",
+                             "Introspect", &sd_bus_err, &m, "");
+  if (r < 0) {
+    ERROR("Can't introspect %s: %s {%s}, %s", unit->path, sd_bus_err.name,
+          sd_bus_err.message, strerror(-r));
+    return EXIT_FAILURE;
+  }
+  char const *xml;
+  r = sd_bus_message_read(m, "s", &xml);
+  xmlDocPtr doc = xmlReadMemory(xml, strlen(xml), "noname.xml", NULL, 0);
```

Comment:
XML is rather inefficient protocol.  Systemd has recently started to support JSON-based varlink interfaces, and those support also introspection: https://www.freedesktop.org/software/systemd/man/255/varlinkctl.html

Is it not possible to get same info through that?